src/api.rs

```diff
@@ -1,10 +1,403 @@
+use std::{fmt, marker::PhantomData};
+
 use crate::{
-    SyntaxKind,
-    cursor::{SyntaxNode, SyntaxToken},
+    cursor, GreenNode, GreenToken, SmolStr, SyntaxText, TextRange, TextUnit, TokenAtOffset,
+    WalkEvent,
 };
 
-impl SyntaxNode {
-    pub fn ancestors(&self) -> impl Iterator<Item = SyntaxNode> {
-        std::iter::successors(Some(self.clone()), SyntaxNode::parent)
+pub trait Language: Sized + Clone + Copy + fmt::Debug + Eq + Ord + std::hash::Hash {
+    fn debug_kind(kind: &SyntaxKind<Self>, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&kind.raw, f)
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct SyntaxKind<L> {
```

Comment:
I thought we agreed on [using traits which the user could implement themselves?](https://gitlab.com/jD91mZM2/rnix/merge_requests/10#note_175424408)

Reply:
Yeah, I've tried that approach, and it turned out to be pretty unwieldy. 

The biggest problem is that the trait-based approach needs a swamp of associated types, and associated types don't work with `impl Trait` (so, using `std::iter::successors` becomes impossible) and they also don't map nicely to `SyntaxElement` enum.

The second problem is that you can't actually use the trait approach to write blanket impls for Debug & Display. You can supply an utility function `debug_fmt`, but the user will have to write an `impl`.

The third problem is that the user will need to import API traits to use them, instead of all methods being inherent.

So I want to try this "non-template base class" pattern, but via composition, and I want to try to migrate rust-analyzer to it, instead of manually doing forwarding there. I *think* my desire to be able to write trait impls is an instance of YAGNI

Reply:
I see, thanks for the insight! :+1: for this way, then

Reply:
also cc https://users.rust-lang.org/t/api-design-question-how-to-wrap-a-family-of-types/30424

Reply:
Crazy idea: A macro to generate code similar to Rust analyzer, that wraps a SyntaxNode. Not using traits will allow you to use `impl Trait`. It won't be pretty, but... I guess it'll let users implement their own `Debug`...........

Reply:
Or, what about abusing that root data field? Since you're already using dynamic dispatch to avoid generics, what if we change `Option<Box<dyn Any + Send + Sync>>` to `Box<dyn Lang>` where
```rust
trait Lang: Any + Send + Sync {
    fn debug_kind(...)
}
// To store root data, one now adds a field to their own struct implementing `Lang`
```

Reply:
Root data is gone in this version :) But yeah, I imagine using dynamic dispatch here *might* work